### PR TITLE
Add `on_success` method

### LINKED ISSUE #1: Add .on_success to Future object
Add an `on_success` subscriber method.  Should work like `on_complete` but only trigger the callback if the resulting `Future` is successful.

### PR DESCRIPTION
**Pull request by sloboegen**
Closes #1

Hello! First, thanks for the interesting project :) I would like to contribute a little.

Some points:

- I copied the code from `on_complete` function, because I didn't want to separate `acquire`/`release` between different methods or do `acquire` twice;
- I dropped py36 from `tox.ini`. If you don't mind, I can make another PR with bumping the python version to 3.11.

### CODE REVIEW

**Comment by vickumar1981** [comment]
@sloboegen   thanks for this!  will take a look this weekend and get it merged in.

**Comment by vickumar1981** [comment]
@sloboegen 

I know you didn't create this bug, and I did, but would you mind changing the name of the class to `TestFuture` in the `test_future.py` file?

[https://github.com/sloboegen/pyeffects/blob/future-on-success/tests/test_future.py#L7](https://github.com/sloboegen/pyeffects/blob/future-on-success/tests/test_future.py#L7)

Thanks

**Comment by vickumar1981** [comment]
@sloboegen , it looks like you bumped the python version in the tests, the `tox.ini` file to `3.9`.

https://github.com/sloboegen/pyeffects/blob/future-on-success/tox.ini#L2

B/c this is a library, I didn't want to jump all the way to 3.11 and instead keep it at a minimum python version that could still be used for versions higher up.

Can we bump this to 3.8 instead?

There's also two other places to make a change for the version, in the Pipfile:

https://github.com/vickumar1981/pyeffects/blob/master/Pipfile#L26

And in the `setup.py` file:

https://github.com/sloboegen/pyeffects/blob/future-on-success/setup.py#L72

Basically, in all three places, let's just go with 3.8 and higher as the standard.

**Comment by vickumar1981** [comment]
LGTM.  @sloboegen  We can either address the comments in this PR or merge this, and open another PR to address the comments.

Your choice.

Thanks again.  :+1:

File: pyeffects/Future.py [hunk 1]
```diff
@@ -198,5 +198,33 @@ def on_complete(self, subscriber: Callable[[A], None]) -> None:
...
+          >>> val.on_success(lambda v: print(v))
+          Success(25)
+
+          >>> def error():
+          ...   raise RuntimeError()
+          >>> Future.run(error).on_success(lambda _: print(42))
+        """
+        self.semaphore.acquire()
+        if self.is_failure():
+            self.semaphore.release()
```
Comment (vickumar1981) [inline pyeffects/Future.py:219]: one more comment, this block isn't currently tested.  

might want to add one more test (i.e., the unhappy path), where you run an unsuccessful future and then run `on_complete`.

**Comment by sloboegen** [comment]
Thanks for your feedback. I bumped Python version to 3.8 in three places you have mentioned and added a test for unhappy case

File: Pipfile [hunk 2]
```diff
@@ -23,4 +23,4 @@ typing_extensions = "*"
 [packages]
 
 [requires]
-python_version = "3.6"
+python_version = "3.8"
```
Comment (vickumar1981) [inline Pipfile:26]: You need to run the command `make init` to update the `Pipfile.lock` and check that file in.

Other than that, the PR looks great and I can merge it.

Comment (sloboegen) [reply Pipfile:26]: Done. But it leaded to some issues...

File: pyeffects/Future.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 from functools import reduce
 import threading
 
-A = TypeVar('A', covariant=True)
+A = TypeVar('A')
```
Comment (sloboegen) [inline pyeffects/Future.py:16]: Mypy fails if A is covariant, but it should be it, isn't it?
Now mypy runs without failures.

Comment (vickumar1981) [reply pyeffects/Future.py:16]: I think it should covariant be b/c if `Dog` is a type of `Animal`, then we'd want a `Future[Dog]` to be a proper subtype of `Future[Animal]`.

The same error is occurring via `mypy` for the other classes too.

This ties back to the idea that all of these classes extend `Monad` which in turn is a type of `Generic[A]`.

I have a fix, but it involves ignoring two lines.

I'll merge this in and submit a separate PR to fix the linter issues.

The problem with covariant types is that they can't be used as input parameters to functions, b/c those types need to be either contravariant types.

Comment (sloboegen) [reply pyeffects/Future.py:16]: Okay, got it. Thank you.
